java/java.hints/src/org/netbeans/modules/java/hints/jdk/ConvertToNestedRecordPattern.java

```diff
@@ -0,0 +1,261 @@
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.RecordComponentElement;
+import javax.lang.model.element.TypeElement;
+import org.netbeans.api.java.source.CompilationInfo;
+import org.netbeans.api.java.source.TreePathHandle;
+import org.netbeans.api.java.source.CodeStyleUtils;
+import org.netbeans.api.java.source.WorkingCopy;
+import org.netbeans.api.java.queries.CompilerOptionsQuery;
+import org.netbeans.modules.java.hints.errors.Utilities;
+import org.netbeans.spi.editor.hints.ErrorDescription;
+import org.netbeans.spi.editor.hints.Fix;
+import org.netbeans.spi.java.hints.ErrorDescriptionFactory;
+import org.netbeans.spi.java.hints.Hint;
+import org.netbeans.spi.java.hints.HintContext;
+import org.netbeans.spi.java.hints.JavaFix;
+import org.netbeans.spi.java.hints.MatcherUtilities;
+import org.netbeans.spi.java.hints.TriggerTreeKind;
+import org.openide.util.NbBundle;
+
+@NbBundle.Messages({
+    "DN_ConvertToNestedRecordPattern=Convert to nested record pattern",
+    "DESC_ConvertToNestedRecordPattern=Convert to nested record pattern",
+    "ERR_ConvertToNestedRecordPattern=Convert to nested record pattern",
+    "FIX_ConvertToNestedRecordPattern=Convert to nested record pattern"
+})
+@Hint(displayName = "#DN_ConvertToNestedRecordPattern", description = "#DESC_ConvertToNestedRecordPattern", category = "rules15",
+        minSourceVersion = "19")
+/**
+ *
+ * @author mjayan
+ */
+public class ConvertToNestedRecordPattern {
+
+    private static final int RECORD_PATTERN_PREVIEW_JDK_VERSION = 19;
+
+    @TriggerTreeKind(Tree.Kind.DECONSTRUCTION_PATTERN)
+    public static ErrorDescription convertToNestedRecordPattern(HintContext ctx) {
+        if (Utilities.isJDKVersionLower(RECORD_PATTERN_PREVIEW_JDK_VERSION) && !CompilerOptionsQuery.getOptions(ctx.getInfo().getFileObject()).getArguments().contains("--enable-preview")) {
+            return null;
+        }
+        TreePath t = ctx.getPath();
+        if (!t.getParentPath().getLeaf().getKind().equals(Tree.Kind.INSTANCE_OF)) {
+            return null;
+        }
+        Set<String> recordPatternVarSet = new HashSet<>();
+        Map<PatternTree, List<PatternTree>> recordComponentMap = new LinkedHashMap<>();
+        DeconstructionPatternTree recordPattern = (DeconstructionPatternTree) t.getLeaf();
+        recordComponentMap = findNested(recordPattern, recordComponentMap);
+
+        for (PatternTree p : recordComponentMap.keySet()) {
+            BindingPatternTree bTree = (BindingPatternTree) p;
+            recordPatternVarSet.add(bTree.getVariable().getName().toString());
+        }
+        while (t != null && t.getLeaf().getKind() != Tree.Kind.IF) {
+            t = t.getParentPath();
+        }
+        Set<TreePath> convertPath = new HashSet<>();
+        List<String> localVarList = new ArrayList<>();
+        Map<String, List<UserVariables>> userVars = new HashMap<>();
+        new TreePathScanner<Void, Void>() {
```

Comment:
Maybe try to use `CancelableTreePathScanner` and try to cancel the process is the hint is canceled (esp. if we are going to search bigger parts of the source code).